From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "guoping.liu" <guoping.liu@3reality.com>
Date: Tue, 17 Mar 2026 10:00:00 +0800
Subject: [PATCH] add arm64 dts ThirdReality linuxbox hub-v3x

V3X is a variant with dual-chip DDR:
- 1024MB RAM (2x 512MB DDR3)
- Old SDK U-Boot 2015.01 with ACS DDR parameter passing
- No Thread UART (ttyAML6 / uart_B disabled)
- uart_AO_B uses uart_ao_b_pins (not uart_ao_b_z_pins)
- Zigbee only (no Thread support)
- ramoops for crash logs, CMA reduced to 4MB

---
 arch/arm64/boot/dts/amlogic/Makefile              |  1 +
 .../amlogic/meson-axg-thirdreality-hub-v3x.dts    |100 ++++++++++++++++++++++
 2 files changed, 101 insertions(+)
 create mode 100644 arch/arm64/boot/dts/amlogic/meson-axg-thirdreality-hub-v3x.dts

diff --git a/arch/arm64/boot/dts/amlogic/Makefile b/arch/arm64/boot/dts/amlogic/Makefile
--- a/arch/arm64/boot/dts/amlogic/Makefile
+++ b/arch/arm64/boot/dts/amlogic/Makefile
@@ -9,6 +9,7 @@
 dtb-$(CONFIG_ARCH_MESON) += meson-axg-thirdreality-hub-v3.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-axg-thirdreality-hub-v3b.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-axg-thirdreality-hub-v3a.dtb
+dtb-$(CONFIG_ARCH_MESON) += meson-axg-thirdreality-hub-v3x.dtb
 # dtb-$(CONFIG_ARCH_MESON) += meson-axg-s400.dtb
 # dtb-$(CONFIG_ARCH_MESON) += meson-g12a-radxa-zero.dtb
 # dtb-$(CONFIG_ARCH_MESON) += meson-g12a-sei510.dtb
diff --git a/arch/arm64/boot/dts/amlogic/meson-axg-thirdreality-hub-v3x.dts b/arch/arm64/boot/dts/amlogic/meson-axg-thirdreality-hub-v3x.dts
new file mode 100644
--- /dev/null
+++ b/arch/arm64/boot/dts/amlogic/meson-axg-thirdreality-hub-v3x.dts
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (c) 2022 Vyacheslav Bocharov <adeep@lexina.in>
+ * Copyright (c) 2022 JetHome
+ * Author: Vyacheslav Bocharov <adeep@lexina.in>
+ */
+
+/dts-v1/;
+
+#include "meson-axg-jethome-jethub-j1xx.dtsi"
+
+/ {
+	compatible = "ThirdReality,trhub-v3x", "amlogic,a113d", "amlogic,meson-axg";
+	model = "ThirdReality TRHub-V3X";
+
+	/* No serial6 (Thread UART) on this variant */
+	aliases {
+		/delete-property/ serial6;
+	};
+
+	/* 1024MB RAM (2x 512MB DDR3) */
+	memory@0 {
+		device_type = "memory";
+		reg = <0x0 0x0 0x0 0x40000000>;
+	};
+
+	reserved-memory {
+		/* Keep secmon as no-map (default from meson-axg.dtsi) */
+
+		linux,cma {
+			size = <0x0 0x400000>;
+		};
+
+		ramoops@0x07400000 {
+			compatible = "ramoops";
+			reg = <0x0 0x07400000 0x0 0x00100000>;
+			record-size = <0x20000>;
+			console-size = <0x40000>;
+			ftrace-size = <0x80000>;
+			pmsg-size = <0x20000>;
+		};
+	};
+
+	/* Amlogic Bluetooth controller */
+	aml_bt: aml_bt {
+		compatible = "amlogic, aml-bt";
+		status = "okay";
+	};
+
+	aml_wifi: aml_wifi {
+		compatible = "amlogic, aml-wifi";
+		status = "okay";
+		irq_trigger_type = "IRQF_TRIGGER_LOW";
+		pwm_config = <&wifi_pwm_conf>;
+	};
+
+	wifi_pwm_conf:wifi_pwm_conf{
+		pwm_channel1_conf {
+			pwms = <&pwm_AO_cd 0 30541 0>;
+			duty-cycle = <15270>;
+			times = <10>;
+		};
+
+		pwm_channel2_conf {
+			pwms = <&pwm_AO_cd 2 30500 0>;
+			duty-cycle = <15250>;
+			times = <12>;
+		};
+	};
+};
+
+/* wifi module */
+&sd_emmc_b {
+	broken-cd;
+};
+
+&aml_wifi{
+	status = "okay";
+	interrupt-gpios = <&gpio  GPIOX_6  GPIO_ACTIVE_HIGH>;
+	power_on-gpios = <&gpio   GPIOX_7  GPIO_ACTIVE_HIGH>;
+};
+
+&aml_bt {
+	status = "okay";
+	reset-gpios = <&gpio    GPIOX_21 GPIO_ACTIVE_HIGH>;
+	btwakeup-gpios = <&gpio GPIOZ_7 GPIO_ACTIVE_HIGH>;
+	hostwake-gpios = <&gpio GPIOZ_6 GPIO_ACTIVE_HIGH>;
+};
+
+/* V3X: uart_AO_B uses uart_ao_b_pins (not uart_ao_b_z_pins) */
+&uart_AO_B {
+	status = "okay";
+	pinctrl-0 = <&uart_ao_b_pins>;
+	pinctrl-names = "default";
+};
+
+/* No Thread UART on V3X */
+&uart_B {
+	status = "disabled";
+};
